tutorial/7 - utils.test.ts

```diff
@@ -0,0 +1,197 @@
+     * the current and the previous state.
+     *
+     * *Note functions like `pairwise` and `scan` can be used with any callback,
+     * so it can be used both in a `.derive()` step and in a `.react()`*
+     */
+    it('pairwise', () => {
+        const myCounter$ = atom(1);
+        const reactSpy = jest.fn();
+
+        /**
+         * ** Your Turn **
+         *
+         * Now, use `pairwise()`, to subtract the previous value from the
+         * current.
+         *
+         * *Hint: check the overloads of pairwise if you're struggling with
+         * `oldVal`.*
+         */
+        myCounter$.derive(__YOUR_TURN__).react(reactSpy);
+
+        expect(reactSpy).toHaveBeenCalledExactlyOnceWith(1, expect.toBeFunction());
+
+        myCounter$.set(3);
+
+        expect(reactSpy).toHaveBeenCalledTimes(2);
+        expect(reactSpy).toHaveBeenCalledWith(2, expect.toBeFunction());
+
+        myCounter$.set(45);
+
+        expect(reactSpy).toHaveBeenCalledTimes(3);
+        expect(reactSpy).toHaveBeenCalledWith(42, expect.toBeFunction());
+    });
+
+    /**
+     * `scan` is the `Derivable` version of `Array.prototype.reduce`. It will be
+     * called with the current state and the last emitted value.
+     *
+     * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)
+     *
+     * *Note: as with `pairwise()` this is useable in both a `.derive()` and
+     * `.react()` method*
+     */
+    it('scan', () => {
+        const myCounter$ = atom(1);
+        const reactSpy = jest.fn();
+
+        /**
+         * ** Your Turn **
+         *
+         * Now, use `scan()`, to add all the emitted values together
+         */
+        myCounter$.derive(__YOUR_TURN__).react(reactSpy);
+
+        expect(reactSpy).toHaveBeenCalledExactlyOnceWith(1, expect.toBeFunction());
+
+        myCounter$.set(3);
+
+        expect(reactSpy).toHaveBeenCalledTimes(2);
+        expect(reactSpy).toHaveBeenCalledWith(4, expect.toBeFunction());
+
+        myCounter$.set(45);
+
+        expect(reactSpy).toHaveBeenCalledTimes(3);
+        expect(reactSpy).toHaveBeenCalledWith(49, expect.toBeFunction());
+
+        /**
+         * *BONUS: Try using `scan()` (or `pairwise()`) directly in the
+         * `.react()` method.*
+         */
+    });
+
+    it.skip('pairwise - BONUS', () => {
+        const myCounter$ = atom(1);
+        let lastPairwiseResult = 0;
+
+        /**
+         * ** Your Turn **
+         * ** BONUS **
+         *
+         * Now, use `pairwise()` directly in `.react()`. Implement the same
+         * derivation as before: subtract the previous value from the current.
+         *
+         * Instead of returning the computed value, assign it
+         * `lastPairwiseResult` instead. This is so the implementation can be
+         * validated.
+         */
+        myCounter$.react(__YOUR_TURN__);
+
+        expect(lastPairwiseResult).toEqual(1);
+
```

Comment:
Dit is denk ik de netste implementatie die ik kan bedenken. Zo schrijf je je code direct in de react, en heb je geen externe variabele.
Deze manier van kijken wat een `reactSpy` returned is common, dus lijkt me net.